### ISSUE TITLE
Trouble setting http basic auth

### ISSUE THREAD

**Issue by pedro-gutierrez** [jcmoraisjr/haproxy-ingress#102]
Hi, 

I am having trouble to make http basic auth work for my ingress.  

I believe I have properly setup my secret and my ingress annotations, according to the docs. As a matter of fact, requests are routed to my backend, and indeed the path prefix is stripped, as of my rewrite rule, however, no matter what I do, basic auth challenge never seems to get triggered.

I inspected the generated haproxy.cfg inside my ingress controller pod, and I noticed the txn.path variable is used before it is actually declared, see this snippet:

```
frontend httpfront-...
    ...
    http-request auth realm ... if { var(txn.path) -m ... } !{ http_auth(...)}
    ...
    http-request set-var(txn.path) path
    ...
    use_backend ... if { var(txn.path) -m ... }
...
```

(I omitted the specifics related to my setup)

I am far from an expert when it comes to HAProxy configuration, however I was wondering whether this could be an actual bug, or a simply user error from my side? As a side note, I have succesfully setup http basic auth with other ingress controllers, such as traefik and nginx.

Anyway, If you need further info please let me know. I'll be more than happy to provide with as much detail as needed in order to get this sorted out.

Thanks in advance,
Pedro.

**Comment by pedro-gutierrez** [comment]
Ok I just realized the behavior I described above applies only to  the canary tag. The stable version works as expected.

**Comment by pedro-gutierrez** [comment]
I submitted this PR https://github.com/jcmoraisjr/haproxy-ingress/pull/103

**Comment by jcmoraisjr** [comment]
Nice catch, thanks! I've just updated my tests with scenarios like this one. I updated `master` and `release-0.5` - the later will be used to update `canary` and shotly `latest`. Let me know if this works for you. If you cannot rebuild from sources I can create a new beta before your tests.

**Comment by jcmoraisjr** [comment]
Just released a new beta with this change, tags `v0.5-beta.2` and `canary`.

**Comment by pedro-gutierrez** [comment]
Hi, thanks a lot for your feedback and for the new tag! It is working very well according to my regression suite. Cheers!

**Comment by jcmoraisjr** [comment]
Thank you for your feedback! Closing.

**Comment by pedro-gutierrez** [comment]
Hi again, actually, not sure this was happening before, but http basic auth stops being triggered as soon as I my secret is re-created. I need to re-deploy my haproxy controller and then things work again. Is this a known issue?

**Comment by jcmoraisjr** [comment]
Couldn't reproduce this on ing v0.5-beta-2 and k8s 1.9.2. What I did:

* created a secret 'mypasswd'
* created an ingress with `ingress.kubernetes.io/auth-secret: mypasswd` and also auth-type and auth-realm
* Edited the secret with `kubectl edit secret mypasswd` base64 encoding the `usr:pwd`

The change was immediate and server was restarted. Change was applied. Please provide some steps and k8s version in order to reproduce this on my env.

**Comment by pedro-gutierrez** [comment]
I was able to reproduce like this. First of all, here is my kubeclt version and ingress:

```
# kubectl --version
Kubernetes v1.9.2+coreos.0
```
```
# kubectl describe ingress nimg -n nc
Name:             nimg
Namespace:        nc
Address:
Default backend:  default-http-backend:80 (<none>)
TLS:
  nc-cert terminates k.netc.io
Rules:
  Host       Path  Backends
  ----       ----  --------
  k.netc.io
             /image   nimg:3000 (<none>)
Annotations:
  auth-realm:      netcomposer
  auth-secret:     ncauth
  auth-type:       basic
  rewrite-target:  /
  ssl-redirect:    true
Events:
  Type    Reason  Age   From                Message
  ----    ------  ----  ----                -------
  Normal  CREATE  5m    ingress-controller  Ingress nc/nimg
  Normal  UPDATE  4m    ingress-controller  Ingress nc/nimg
```

I created a password hash as follows:

```
 # mkpasswd -m sha-512 -s test
$6$fieFgYTbe0zr$FxZy/cZfs.FvRBeC.RSjRVbkGYZCCI9NIPyKcUbrHbhnzBDJ1m401wqH9uy9pCJ8sWtN6yLnAzKtI/H0gnyl./
```

and this is what my `auth` file looks like:

```
# cat auth
john:$6$fieFgYTbe0zr$FxZy/cZfs.FvRBeC.RSjRVbkGYZCCI9NIPyKcUbrHbhnzBDJ1m401wqH9uy9pCJ8sWtN6yLnAzKtI/H0gnyl./
```

With this setup, and after a fresh deployment of ingress, I can verify the following works well:

```
# curl https://k.netc.io/image/
<html><body><h1>401 Unauthorized</h1>
You need a valid user and password to access this content.
</body></html>
```

```
root@node1:~/automation# curl https://k.netc.io/image/ -u john:test
{"memory":{"rss":61198336,"heapTotal":24158208,"heapUsed":19586968,"external":1487708},"sharp":{"versions":{"cairo":"1.14.12","croco":"0.6.12","exif":"0.6.21","expat":"2.2.5","ffi":"3.2.1","fontconfig":"2.12.6","freetype":"2.9","gdkpixbuf":"2.36.11","gif":"5.1.4","glib":"2.55.1","gsf":"1.14.42","harfbuzz":"1.7.4","jpeg":"1.5.3","lcms":"2.9-","orc":"0.4.28","pango":"1.41.0","pixman":"0.34.0","png":"1.6.34","svg":"2.42.0","tiff":"4.0.9-cda4b06","vips":"8.6.1","webp":"0.6.1","xml":"2.9.7","zlib":"1.2.11"},"counters":{"queue":0,"process":0},"cache":{"memory":{"current":0,"high":0,"max":50},"files":{"current":0,"max":20},"items":{"current":0,"max":100}},"concurreny":1,"simd":true}}
```

Now, if I delete my secret:

```
# kubectl delete secret ncauth -n nc
secret "ncauth" deleted
```

Then I notice http basic auth no longer gets triggered, and my url is now un-protected:

```
# curl https://k.netc.io/image/
{"memory":{"rss":109699072,"heapTotal":26779648,"heapUsed":23405056,"external":2278668},"sharp":{"versions":{"cairo":"1.14.12","croco":"0.6.12","exif":"0.6.21","expat":"2.2.5","ffi":"3.2.1","fontconfig":"2.12.6","freetype":"2.9","gdkpixbuf":"2.36.11","gif":"5.1.4","glib":"2.55.1","gsf":"1.14.42","harfbuzz":"1.7.4","jpeg":"1.5.3","lcms":"2.9-","orc":"0.4.28","pango":"1.41.0","pixman":"0.34.0","png":"1.6.34","svg":"2.42.0","tiff":"4.0.9-cda4b06","vips":"8.6.1","webp":"0.6.1","xml":"2.9.7","zlib":"1.2.11"},"counters":{"queue":0,"process":0},"cache":{"memory":{"current":3,"high":19,"max":50},"files":{"current":0,"max":20},"items":{"current":100,"max":100}},"concurreny":1,"simd":true}}
```

And if I create my secret again:

```
# kubectl create secret generic ncauth --from-file auth --namespace nc
secret "ncauth" created
```

that does not seem to fix the problem either, my url is still unprotected.

```
# curl https://k.netc.io/image/
{"memory":{"rss":109699072,"heapTotal":26779648,"heapUsed":23405056,"external":2278668},"sharp":{"versions":{"cairo":"1.14.12","croco":"0.6.12","exif":"0.6.21","expat":"2.2.5","ffi":"3.2.1","fontconfig":"2.12.6","freetype":"2.9","gdkpixbuf":"2.36.11","gif":"5.1.4","glib":"2.55.1","gsf":"1.14.42","harfbuzz":"1.7.4","jpeg":"1.5.3","lcms":"2.9-","orc":"0.4.28","pango":"1.41.0","pixman":"0.34.0","png":"1.6.34","svg":"2.42.0","tiff":"4.0.9-cda4b06","vips":"8.6.1","webp":"0.6.1","xml":"2.9.7","zlib":"1.2.11"},"counters":{"queue":0,"process":0},"cache":{"memory":{"current":3,"high":19,"max":50},"files":{"current":0,"max":20},"items":{"current":100,"max":100}},"concurreny":1,"simd":true}}
```

My only workaround is to re-deploy my ingress: 

```
# kubectl delete ingress nimg -n nc
ingress "nimg" deleted
```
```
# kubectl create -f nimg-ingress.yml
ingress "nimg" created
```

and only then things are back to normal:

```
# curl https://k.netc.io/image/
<html><body><h1>401 Unauthorized</h1>
You need a valid user and password to access this content.
</body></html>
```

**Comment by jcmoraisjr** [comment]
Yup, you deleted and recreated the secret while I changed without delete. Please check if the canary tag solved this another issue.

**Comment by pedro-gutierrez** [comment]
Hey there, with the latest canary tag, I can see the following behavior:

1. when I delete the secret, then the url  for my ingress is exposed  and available without http basic auth.
2. when I re-create the secret, then the url is again protected with basic auth.

So, as far as 2. is concerned, the latest canary tag is working better now. Thank you for the fix! 

Now, is there anything that can be done in order to address 1. ? Actually, I am not quite sure what would be the behavior to expect, when an ingress defines the `auth-secret` annotation, but the associated secret does not exist?

**Comment by jcmoraisjr** [comment]
This was already addressed on #71 but wasn't added on v0.5 - which should be tagged 2 months ago, explaining why that patch wasn't there.

Closing this issue.

**Comment by pedro-gutierrez** [comment]
Perfect! Thank u!
